README
NAME
            Acme::Class::DBI::NoThreats - Simple Database Abstraction

SYNOPSIS
            package Music::DBI;
            use base 'Acme::Class::DBI::NoThreats';
            Music::DBI->set_db('Main', 'dbi:mysql', 'username', 'password');

            package Artist;
            use base 'Music::DBI';
            Artist->table('artist');
            Artist->columns(All => qw/artistid name/);
            Artist->has_many('cds', 'CD' => artist);

            package CD;
            use base 'Music::DBI';
            CD->table('cd');
            CD->columns(All => qw/cdid artist title year/);
            CD->has_many('tracks', 'Track' => 'cd', { sort => 'position' });
            CD->has_a(artist => 'CD::Artist');
            CD->has_a(reldate => 'Time::Piece',
                    inflate => sub { Time::Piece->strptime(shift => "%Y-%m-%d") },
                    deflate => 'ymd',
            }

            CD->might_have(liner_notes => LinerNotes => qw/notes/);

            package Track;
            use base 'Music::DBI';
            Track->table('track');
            Track->columns(All => qw/trackid cd position title/); 

            #-- Meanwhile, in a nearby piece of code! --#

            my $artist = Artist->create({ artistid => 1, name => 'U2' });

            my $cd = $artist->add_to_cds({ 
                    cdid   => 1,
                    title  => 'October',
                    year   => 1980,
            });

            # Oops, got it wrong.
            $cd->year(1981);
            $cd->commit;

            # etc.

            while (my $track = $cd->tracks) {
                    print $track->position, $track->title
            }

            $cd->delete; # also deletes the tracks

            my $cd  = CD->retrieve(1);
            my @cds = CD->retrieve_all;
            my @cds = CD->search(year => 1980);
            my @cds = CD->search_like(title => 'October%');

PREFACE
    Acme::Class::DBI::NoThreats is a fork of Class::DBI version 0.96. It is
    identical in every aspect, except for the package name, and that the
    current maintainer agrees NOT TO:

    *   Threaten to sue users who subscribe to the mailing list, or to call
        the police.

    *   threaten to sue users who quote the maintainer's opinions.

    *   threaten to sue users who reconstruct wikis or other materials
        related to this project.

    *   threaten to sue users who post opinions about this module.

    *   threaten to sue users who think things that the maintainer finds
        disagreeable.

    Note that, other than that, THERE ARE NO FURTHER GUARANTEES.
    Specifically, there's no guarantee that the maintainer of the original,
    pre-fork Class::DBI will abide by these terms. There's no guarantee that
    Class::DBI plugins will work. And there's also no guarantee that this
    forked version will be maintained at all! (It's "ACME::", after all.)

DESCRIPTION
    Acme::Class::DBI::NoThreats provides a convenient abstraction layer to a
    database.

    It not only provides a simple database to object mapping layer, but can
    be used to implement several higher order database functions (triggers,
    referential integrity, cascading delete etc.), at the application level,
    rather than at the database.

    This is particularly useful when using a database which doesn't support
    these (such as MySQL), or when you would like your code to be portable
    across multiple databases which might implement these things in
    different ways.

    In short, Acme::Class::DBI::NoThreats aims to make it simple to introduce
    'best practice' when dealing with data stored in a relational database.

PRE-REQUISITES

        Acme::Class::DBI::NoThreats requires on several other modules.
        Installing via the CPAN shell should take care of all these for you.
        But if you need to install by hand you need to have the following:

	To run the module:

 		Class::Accessor          
		Class::Data::Inheritable
		Class::Trigger         
 		Ima::DBI
			(which in turn requires DBI itself, and Class::WhiteHole)

    To test the installation:
       Test::More
       DBD::SQLite
       DBD::mysql
       DBD::Pg


Changes
0.96 forked version
    - New namespace: Acme::Class::DBI::NoThreats

0.96

  New Functionality
    - Maintain live object index so subsequent requests for same row return 
      the same object (Perrin Harkins)
    - connection() is now preferred over set_db()
    - New relationship architecture (see Class::DBI::Relationship)
    - meta_info now returns more information about relationships 
    - search($column => undef) now does an ISNULL search (Bart Lateur,
      Drew Wilson, William McKee)
    - A slice of an iterator in scalar context now gives another
      iterator (Thomas Klausner)
    - Attribute methods can now take slices (Charles Bailey)
    - pass $self to inflation/deflation methods, allowing (amongst other
      things, cross-column processing in a custom method) (Charles Bailey)
    - deflation now happens at distinct trigger points that happen after
      normal before_create/update triggers (Dave Cash)
    - Columns may now set their placeholders:
        Foo->find_column('tdate')->placeholder("IF(1, CURDATE(), ?)");
    - Columns know if they have been constrained

  Documentation
    - add note about needing a RootClass (and fix tests for this)
    - Explain interaction of is_changed and auto-update (William McKee)
    - Better examples for select_val and sql_single (Paul Makepeace)
    - Add note about sprintf interpolation and wildcard searches in
      home-grown queries (Christopher L. Everett)
    - Mention the wiki

  Bug Fixes
    - Better error reporting from failed inflations [Tatsuhiko Miyagawa]
    - You can now have a column named 'first' [Tatsuhiko Miyagawa]
    - inflation and deflation handle edge cases better [Charles Bailey]
    - search_ methods are only generated for SELECT calls
    - add_constructor is now propery subclassable (Thomas Klausner)

  Deprecations
    - Class::DBI::Query is now deprecated. The SQL substitutions are a
      much better approach.
    - sort argument to has_many is now order_by to match search
    - __hasa_rels and __hasa_list are deprecated in favour of meta_info

  Other
    - construct() is no longer a protected method. 
    - skip t/17 on error. There's a bug here I can't find, or fix, yet.

0.95 Fri Jan  9 2004

  Pre-Requisites
    - Now requires Perl 5.6 or greater

  New Functionality
    - constrain_column() adds simple per-column constraints
    - has_many now attempts to work out the reference column by
      examining the foreign class's has_a() declarations, making the 3rd
      argument to has_many mostly obsolete.
    - __TABLE(Other::Class)__ now interpolates the table from the other
      class, allowing avoidance of hard-coding other table names.
    - __ESSENTIAL(alias)__ now prepends the table alias supplied to the
      start of each column name, for avoiding column name clashes.
    - __JOIN(alias alias)__ now inserts the SQL to join the two tables.
    - might_have now allows creating the remote object through a simple
      set of the accessor. [David Jack Olrik]
    - Iterators can now be reset() [Tatsuhiko Miyagawa]

  Documentation
    - count_all, maximum_value_of and minimum_value_of are now
      documented (along with how to implement similar methods)

  Bug Fixes
    - Allow retrieve() to take unnormalised columns, as with search [Connie]
    - Handle stringification before Primary Key(s) are filled 
    - search() can take an overridden accessor name as a key
    - MultiColumn primary keys working better with ColumnGrouper [Perrin
      Hawkins, Paul Makepeace, and Charles Bailey]

  Deprecations
    - _single_value_select and _single_row_select deprecated in favour
      of new select_val and select_row methods in Ima::DBI

  Other
    - Class::Accessor can no longer be used directly to add other attributes.
      TEMP columns should be used for this.
    - Ensure all croaks internally go through _croak() [Richard Piacentini]
    - deleted objects now blessed into Class::DBI::Object::Has::Been::Deleted
    - Lots of doc patches [Jesse Sheidlower, Paul Makepeace]
    - All internal attribute accesses now go through helper methods,
      as first step to changing how all this works.
    - set_sql now passes all arguments up to Ima::DBI (so you can turn
      off caching if required) [gfalck]
    - We now warn about column names clashing with any inherited methods,
      not just ones defined in Class::DBI itself [Dan Thill]
    - Silence some test warnings reported by Paul Makepeace

0.94  Wed Aug 27 2003

  New Functionality
    - allow has_a() columns to hold NULL values (Dominic Mitchell)
    - cascading deletions can be turned off by passing argument to
      has_many setup: no_cascade_delete => 1 (undocumented/untested)
      [now implemented as a trigger]
    - classes and objects can be marked as read_only (experimental and
      undocumented, although tested)

  Documentation
    - no longer refer to Class::DBI::Join, which is dead
    - warn about inflating a primary key using has_a
    - rewrite of docs on transactions
    - numerous small tweaks from Jay Strauss, Alexander Karelas and Brad
      Bowman
    - redocument dbi_commit and dbi_rollback that have vanished...

  Bug Fixes
    - make sequences work again (Dominic Mitchell) [now with tests!]
    - interpolation of __TABLE__ etc now happens globally, rather than
      only once [Corion]
    - might_have now properly does a cascading delete
    - Fixed problem where we can't update an object which might_have a
      relationship, but doesn't.
    - might_have's tentative, but broken, support for MCPK has been
      removed.
    - Increase Class::Accessor pre-requisite to 0.18 to fix strange bug
      reported by Casey West.
    - Make sure we don't alter the hashref being passed to create()
    - The after_create() trigger is now called after the object is
      emptied, and will thus be able to see objects in has_a columns
      (by reloading from the DB) [If you were using this

Makefile.PL

use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

$PACKAGE = 'Acme::Class::DBI::NoThreats';
($PACKAGE_FILE) = $PACKAGE =~ /::([^:]+)/;
$LAST_API_CHANGE = 0.93;

require 5.006;

eval "require $PACKAGE";

unless ($@) { # Make sure we did find the module.
    print <<"CHANGE_WARN" if ${$PACKAGE.'::VERSION'} < $LAST_API_CHANGE;

NOTE: There have been API changes between this version and any older than
version $LAST_API_CHANGE!  Please check carefully if you are upgrading
from a version older than $LAST_API_CHANGE.

Significant changes:

  0.95 - Breaking the encapsulation of an object (e.g. assuming that 
         the value for the 'title' column will be stored internally in
         $object->{'title'} will stop working in a forthcoming release.
         Relying on this behaviour is a Bad Thing, and you should fix it 
         now! (This also applies to before_create triggers, where 
         previously it was the supported approach.)

  0.93 - Undocumented ordered_search() method no longer exists
       - Undocumented make_filter() method no longer exists
       - Undocumented option for add_constructor to do %s subitutions removed
       - single_value_select() no longer takes raw SQL fragments

  0.91 - on_setting() trigger point replaced by new individual column
          triggers (see documentation on constraints)
       - runtime database errors now throw exceptions
         - all errors standardised so on_failed_create no longer needed
       - objects now overload in string or boolean context

  0.90 - hasa is now deprecated in favour of has_a. Therefore
          has_many no longer creates reciprocal hasa relationship.

  0.85 - Acme::Acme::Class::DBI::NoThreats::NoThreats no longer supports pseudo-hash based objects
       - hasa_list is now deprecated in favour of has_many
       - has_many auto-creates a reciprocal hasa relationship

  0.32 - delete() now removes any foreign elements, to avoid orphans

CHANGE_WARN
}

sub MY::postamble { # DBD_PG_USER=tony make cover
	return <<''
cover:
	rm -rf cover_db
	PERL5OPT=-MDevel::Cover \$(MAKE) test || true
	cover cover_db -report html

}

WriteMakefile(
	NAME                => $PACKAGE,
	VERSION_FROM        => 'lib/Acme/Class/DBI/NoThreats.pm',
	PREREQ_PM           => {
		'Class::Accessor'          => '0.18',
		'Class::Data::Inheritable' => '0.02',
		'Class::Trigger'           => '0.07',
		'File::Temp'               => '0.12',
		'Ima::DBI'                 => '0.33',
		'List::Util'               => '1.00',
		'Storable'                 => '2.04',
		'Test::More'               => '0.47',
		'UNIVERSAL::moniker'       => '0.06',
  },
	dist              => { 
		COMPRESS   => 'gzip -9',
		SUFFIX     => '.gz',
		DIST_DEFAULT   => 'all tardist',
	},
);
